flow/workflows/maintenance_flow.go

```diff
@@ -0,0 +1,275 @@
+		StartToCloseTimeout: 24 * time.Hour,
+	})
+
+	snapshotWaitCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 24 * time.Hour,
+		HeartbeatTimeout:    1 * time.Minute,
+	})
+	waitSnapshotsFuture := workflow.ExecuteActivity(snapshotWaitCtx,
+		maintenance.WaitForRunningSnapshots,
+	)
+	err := waitSnapshotsFuture.Get(snapshotWaitCtx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	enableMaintenanceFuture := workflow.ExecuteActivity(ctx, maintenance.EnableMaintenanceMode)
+	err = enableMaintenanceFuture.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Waiting for all snapshot mirrors to finish snapshotting")
+	waitSnapshotsPostEnableFuture := workflow.ExecuteActivity(snapshotWaitCtx,
+		maintenance.WaitForRunningSnapshots,
+	)
+	err = waitSnapshotsPostEnableFuture.Get(snapshotWaitCtx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	mirrorsList, err := getAllMirrors(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	runningMirrors, err := pauseAndGetRunningMirrors(ctx, mirrorsList, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	future := workflow.ExecuteActivity(ctx, maintenance.BackupAllPreviouslyRunningFlows, runningMirrors)
+	err = future.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("StartMaintenance workflow completed")
+	return &protos.StartMaintenanceFlowOutput{
+		Version: peerdbenv.PeerDBVersionShaShort(),
+	}, nil
+}
+
+func pauseAndGetRunningMirrors(
+	ctx workflow.Context,
+	mirrorsList activities.MaintenanceMirrorsInfo,
+	logger log.Logger,
+) (activities.MaintenanceMirrorsInfo, error) {
+	selector := workflow.NewSelector(ctx)
+	runningMirrors := make([]bool, len(mirrorsList.Mirrors))
+	for i, mirror := range mirrorsList.Mirrors {
+		activityInput := mirror
+		f := workflow.ExecuteActivity(
+			ctx,
+			maintenance.PauseMirrorIfRunning,
+			activityInput,
+		)
+
+		selector.AddFuture(f, func(f workflow.Future) {
+			var wasRunning bool
+			err := f.Get(ctx, &wasRunning)
+			if err != nil {
+				logger.Error("Error checking and pausing mirror", "mirror", mirror, "error", err)
+			} else {
+				logger.Info("Finished check and pause for mirror", "mirror", mirror, "wasRunning", wasRunning)
+				runningMirrors[i] = wasRunning
+			}
+			ctx.Done()
+		})
+	}
+	onlyRunningMirrors := make([]activities.MaintenanceMirrorInfoItem, 0)
+	for range mirrorsList.Mirrors {
+		selector.Select(ctx)
+		if err := ctx.Err(); err != nil {
+			return activities.MaintenanceMirrorsInfo{}, err
+		}
+	}
+	for i, mirror := range mirrorsList.Mirrors {
+		if runningMirrors[i] {
+			onlyRunningMirrors = append(onlyRunningMirrors, mirror)
+		}
+	}
+	return activities.MaintenanceMirrorsInfo{
+		Mirrors: onlyRunningMirrors,
+	}, nil
+}
+
+func getAllMirrors(ctx workflow.Context) (activities.MaintenanceMirrorsInfo, error) {
+	getMirrorsFuture := workflow.ExecuteActivity(ctx, maintenance.GetAllMirrors)
+	var mirrorsList activities.MaintenanceMirrorsInfo
```

Comment:
we usually make activity results protobuf since temporal has special support for protobuf & helps avoid compatibility issues between versions